File: go-cli/internal/protocol/protocol.go
// PRD 4장 프로토콜 스펙 v1의 공통 상수·스키마 (Dart core와 와이어 호환이 계약)
package protocol

import (
	"encoding/json"
	"fmt"
	"time"
)

const (
	App              = "easy-send"
	Version          = "1.0"
	MulticastGroup   = "224.0.0.168"
	DiscoveryPort    = 53318
	ServicePort      = 53318
	APIPrefix        = "/api/v1"
	AnnounceInterval = 5 * time.Second
	DeviceTTL        = 15 * time.Second
	ApprovalTimeout  = 60 * time.Second
)

// DeviceInfo = announce 스키마에서 announce 플래그를 뺀 기기 정보 (4.2)
type DeviceInfo struct {
	App         string `json:"app"`
	Version     string `json:"version"`
	Alias       string `json:"alias"`
	DeviceType  string `json:"deviceType"`
	Fingerprint string `json:"fingerprint"`
	Port        int    `json:"port"`
	Protocol    string `json:"protocol"`
}

type Announcement struct {
	DeviceInfo
	Announce bool `json:"announce"`
}

// 다른 앱이거나 메이저 버전이 다르면 무시한다 (4.2 필드 규칙)
func (d DeviceInfo) Compatible() bool {
	return d.App == App && majorOf(d.Version) == majorOf(Version)
}

func majorOf(v string) string {
	for i := 0; i < len(v); i++ {
		if v[i] == '.' {
			return v[:i]
		}
	}
	return v
}

func DecodeAnnouncement(data []byte) (Announcement, error) {
	var a Announcement
	if err := json.Unmarshal(data, &a); err != nil {
		return a, err
	}
	if a.Compatible() == false {
		return a, fmt.Errorf("호환되지 않는 피어: app=%q version=%q", a.App, a.Version)
	}
	return a, nil
}

// FileMeta = prepare-upload files 항목 (4.3)
type FileMeta struct {
	Name string `json:"name"`
	Size int64  `json:"size"`
	Mime string `json:"mime"`
}

type PrepareUploadRequest struct {
	Info  DeviceInfo          `json:"info"`
	Files map[string]FileMeta `json:"files"`
}

type PrepareUploadResponse struct {
	SessionID string            `json:"sessionId"`
	Tokens    map[string]string `json:"tokens"`
}

File: app/rust/src/core/protocol.rs
// PRD 4장 프로토콜 스펙 v1의 공통 상수·스키마 (Dart core·Go CLI와 와이어 호환이 계약)
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::net::Ipv4Addr;
use std::time::Duration;

pub const APP: &str = "easy-send";
pub const VERSION: &str = "1.0";
pub const MULTICAST_GROUP: Ipv4Addr = Ipv4Addr::new(224, 0, 0, 168);
pub const DISCOVERY_PORT: u16 = 53318;
pub const SERVICE_PORT: u16 = 53318;
pub const API_PREFIX: &str = "/api/v1";
pub const ANNOUNCE_INTERVAL: Duration = Duration::from_secs(5);
pub const DEVICE_TTL: Duration = Duration::from_secs(15);
pub const APPROVAL_TIMEOUT: Duration = Duration::from_secs(60);

// announce 스키마의 기기 부분 (4.2). Dart 피어는 /info·prepare-upload의 info에
// 이 4필드만 보내므로 app/version/protocol은 여기 두지 않는다.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PeerInfo {
    pub alias: String,
    pub device_type: String,
    pub fingerprint: String,
    pub port: u16,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Announcement {
    pub app: String,
    pub version: String,
    #[serde(flatten)]
    pub info: PeerInfo,
    #[serde(default = "default_scheme")]
    pub protocol: String,
    pub announce: bool,
}

fn default_scheme() -> String {
    "https".to_string()
}

impl Announcement {
    pub fn new(info: PeerInfo, announce: bool) -> Self {
        Self {
            app: APP.to_string(),
            version: VERSION.to_string(),
            info,
            protocol: default_scheme(),
            announce,
        }
    }

    // 다른 앱이거나 메이저 버전이 다르면 무시한다 (4.2 필드 규칙)
    pub fn compatible(&self) -> bool {
        self.app == APP && major_of(&self.version) == major_of(VERSION)
    }

    pub fn decode(data: &[u8]) -> Option<Announcement> {
        let ann: Announcement = serde_json::from_slice(data).ok()?;
        if ann.compatible() {
            Some(ann)
        } else {
            None
        }
    }
}

fn major_of(version: &str) -> &str {
    version.split('.').next().unwrap_or(version)
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FileMeta {
    pub name: String,
    pub size: u64,
    pub mime: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PrepareUploadRequest {
    pub info: PeerInfo,
    pub files: HashMap<String, FileMeta>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareUploadResponse {
    pub session_id: String,
    pub tokens: HashMap<String, String>,
}
